rfc/0009-policy-report.md

```diff
@@ -0,0 +1,323 @@
+about actions happening inside of it. This is going to be a mix of
+some `ClusterAdmissionPolicy` policies and all the `AdmissionPolicy` policies
+that are defined inside of the Namespace.
+
+Next, this list of policies is going to be filtered: only policies interested
+in `CREATE` events are going to be considered.
+The audit report is not going to simulate `DELETE` and `UPDATE` events.
+
+For each relevant policy, Kubewarden will find all the Kubernetes resources that
+are 1) defined inside of the Namespace and 2) are relevant to the policy. Kubewarden
+will then simulate the `CREATE` event of this resource and keep track of the policy
+evaluation result.
+
+All the results are going to be collected into the `PolicyReport` object defined
+inside of the Namespace. There's going to be only one `PolicyReport` per Namespace.
+
+The `PolicyReport` resource has the following fields:
+
+* `apiVersion` and `kind`: these are defined by the CRDs
+* `metadata`: this is the usual `meta/v1.ObjectMeta` resource. In the
+  beginning we will set only the `name` field. We will do that by using the
+  following pattern: `polr-ns-<namespace name>`
+* `scope`: this is a `core/v1.ObjectReference` resource. We will use that to
+  reference the Kubernetes Namespace resource that has been analyzed
+* `scopeSelector`: we will not use that, because we are using the `scope` field
+* `summary`: this is an object defined by a CRD `PolicyReportSummary`. It's made
+  of the following fields:
+  * `pass`: count of policies whose requirements were met
+  * `fail`: count of policies whose requirements were not met
+  * `warn:`: we will not set it, it doesn't relate to us
+  * `error`: count of policies that could not be evaluated.
+    The policies could not be evaluated because of errors
+    (for example: a runtime panic of a broken policy)
+  * `skip`: count of policies that were not selected for evaluation.
+    We will allow some policies to be excluded from the background checks, more
+    on that later.
+* `results` this probably the most important field. It's a list of `PolicyReportResult`
+  objects
+
+The structure of the `PolicyReportResult` is flexible. This is how we are going
+to use its fields:
+
+* `source`: identifier for the policy engine that manages this report. We will
+  set that to be `kubewarden`
+* `policy`: this is the name of the policy that evaluated the resource. This
+  could be both a `ClusterAdmissionPolicy` or a `AdmissionPolicy`.
+  Because of that we cannot use the name of the policy, there could be clashes
+  between cluster and namespaced policy names. We will instead use the following
+  convention to build the contents of this field: `<[cap|ap]>-<policy name>`.
+  For example, a `ClusterAdmissionPolicy` named `provileged-containers` will
+  be identified via `cap-privileged-containers`. On the other hand, a `AdmissionPolicy`
+  called `required-labels` will be identified via `ap-required-labels`.
```

Comment:
Don't we [add prefixes](https://github.com/kubewarden/kubewarden-controller/blob/1c8afcce20f11f1de5056b8adcda53b51e455b3d/apis/policies/v1/clusteradmissionpolicy_types.go#L175) in the cluster policies to avoid this kind of issue? I think we do not need this in the reports.

Reply:
I think this is an internal string in the controller, and never gets set in the resource itself. But at least, we could adopt the same convention with `clusterwide-`.

Reply:
I think we add the prefix `clusterwide-` or `namespaced-nsname- `to the webhook configutarion and the name in the policies configmap to avoid having a `ClusterAdmissionPolicy` and an `AdmissionPolicy` with the same name.
I would use the same naming convention for `PolicyReports`

Reply:
Yeah, I made a confusion with the webhook and policy names. Sorry. We need the prefix indeed.